cpukit/score/src/schedulerstrongapa.c

```diff
@@ -3,339 +3,820 @@
+      _Scheduler_strong_APA_Allocate_processor
+    );
 
-  return (Scheduler_Node *) _Scheduler_priority_Ready_queue_first(
-    &self->Bit_map,
-    &self->Ready[ 0 ]
-  );
+    curr_node = next_node;
+    curr_strong_node = _Scheduler_strong_APA_Node_downcast( curr_node );
+
+    while ( curr_node !=  lowest_reachable ) {
+      curr_CPU = curr_strong_node->cpu_to_preempt;
+      next_node = _Scheduler_strong_APA_Get_scheduled( self, curr_CPU );
+      /* curr_node preempts the next_node; */
+      _Scheduler_SMP_Preempt(
+	context,
+	curr_node,
+	next_node,
+	_Scheduler_strong_APA_Allocate_processor
+      );
+
+      curr_node = next_node;
+      curr_strong_node = _Scheduler_strong_APA_Node_downcast( curr_node );
+    }
+
+    _Scheduler_strong_APA_Move_from_scheduled_to_ready( context, lowest_reachable );
+
+    needs_help = false;
+  } else {
+    needs_help = true;
+  }
+
+  /* Add it to Ready chain since it is now either scheduled or just ready. */
+  _Scheduler_strong_APA_Insert_ready( context,node, insert_priority );
+
+  return needs_help;
 }
 
-void _Scheduler_strong_APA_Block(
-  const Scheduler_Control *scheduler,
-  Thread_Control          *the_thread,
-  Scheduler_Node          *node
+/*
+ * BFS Algorithm for task arrival
+ * Enqueue node either in the scheduled chain or in the ready chain.
+ * node is the newly arrived node and is currently not scheduled.
+ */
+static inline bool _Scheduler_strong_APA_Enqueue(
+  Scheduler_Context *context,
+  Scheduler_Node    *node,
+  Priority_Control   insert_priority
 )
 {
-  Scheduler_Context *context = _Scheduler_Get_context( scheduler );
+  Scheduler_strong_APA_Context *self;
+  Scheduler_strong_APA_CPU     *CPU;
+  uint32_t	                cpu_max;
+  uint32_t              	cpu_index;
+  Per_CPU_Control              *cpu_to_preempt;
+  Scheduler_Node               *lowest_reachable;
+  Scheduler_strong_APA_Node    *strong_node;
 
-  _Scheduler_SMP_Block(
+  /* Denotes front and rear of the queue */
+  uint32_t	front;
+  uint32_t	rear;
+
+  front = 0;
+  rear = -1;
+
+  self = _Scheduler_strong_APA_Get_self( context );
+  strong_node = _Scheduler_strong_APA_Node_downcast( node );
+  cpu_max = _SMP_Get_processor_maximum();
+  CPU = self->CPU;
+
+  for ( cpu_index = 0 ; cpu_index < cpu_max ; ++cpu_index ) {
+    CPU[ cpu_index ].visited = false;
+
+    /* Checks if the thread_CPU is in the affinity set of the node */
+    if ( _Processor_mask_Is_set( &strong_node->Affinity, cpu_index ) ) {
+      Per_CPU_Control *cpu = _Per_CPU_Get_by_index( cpu_index );
+
+      if ( _Per_CPU_Is_processor_online( cpu ) ) {
+        rear = rear + 1;
+        CPU[ rear ].cpu = cpu;
+        CPU[ cpu_index ].visited = true;
+        CPU[ cpu_index ].preempting_node = node;
+      }
+    }
+  }
+
+  /*
+   * This assert makes sure that there always exist an element in the
+   * Queue when we start the queue traversal.
+   */
+  _Assert( !_Processor_mask_Is_zero( &strong_node->Affinity ) );
```

Comment:
Would it be a bug for that to happen, or is it possible an application somehow specifies this behavior?

Reply:
According to the paper, it should be acceptable. The node(Tnew) would remain unassigned.  So I have made some changes, please check the latest commit. These changes now allow for having a node with empty affinity set. But, having a node with empty affinity set does not make sense, since at the end of Enqueue function, if the newly arrived node is not scheduled, it asks for help - which checks for the lowest scheduled directly reachable node - which cannot be checked if the node has an empty affinity set. Thus it does not matter whether we have an assert here or not. I have removed the assert for now.